Source/DafnyLanguageServer/Handlers/Custom/DafnyCounterExampleHandler.cs

```diff
@@ -16,16 +16,27 @@ namespace Microsoft.Dafny.LanguageServer.Handlers.Custom {
   public class DafnyCounterExampleHandler : ICounterExampleHandler {
     private readonly ILogger logger;
     private readonly IDocumentDatabase documents;
+    private readonly ITextDocumentLoader documentLoader;
 
-    public DafnyCounterExampleHandler(ILogger<DafnyCounterExampleHandler> logger, IDocumentDatabase documents) {
+    public DafnyCounterExampleHandler(ILogger<DafnyCounterExampleHandler> logger, ITextDocumentLoader documentLoader, IDocumentDatabase documents) {
       this.logger = logger;
       this.documents = documents;
+      this.documentLoader = documentLoader;
     }
 
     public async Task<CounterExampleList> Handle(CounterExampleParams request, CancellationToken cancellationToken) {
       try {
         var document = await documents.GetLastDocumentAsync(request.TextDocument);
+
         if (document != null) {
+          var verificationTasks = document!.VerificationTasks;
+          if (verificationTasks != null) {
+            var entry = documents.Documents[request.TextDocument.Uri];
```

Comment:
I don't understand these two levels of nesting: `GetLastDocumentAsync` looks up `TextDocument` using its URI, but then we look up `TextDocument.Uri` again in the result?

Reply:
The type of `entry` is `DocumentEntry`, which is different from `DafnyDocument`. `GetLastDocumentAsync` is an async call that makes sure the document is translated, then later we need `entry` to verify the individual tasks. This all looks better in the refactored PR: https://github.com/dafny-lang/dafny/pull/2478/files#diff-d628271b7999a61822dcdc03e39b669393f7f504799c979c7fa7b1bf8f4911f9R27